### ISSUE TITLE
Modifying the instancetype does not cause the virtual machine's Current CPU Topology value to refresh.

### ISSUE THREAD

**Issue by tghfly** [kubevirt/kubevirt#11251]
**What happened**:
I define an instance type instanceType t1
```yaml
apiVersion: instancetype.kubevirt.io/v1beta1
kind: VirtualMachineInstancetype
metadata:
  name: t1
  namespace: default
spec:
  cpu:
    guest: 1
    model: host-passthrough
  memory:
    guest: 1Gi
  nodeSelector:
    cpu: true
```
> and then create a vm vm-ubuntu03 based on t1
```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: vm-ubuntu03
spec:
  instancetype:
    kind: VirtualMachineInstancetype
    name: t1
  running: true
  template:
    metadata:
      labels:
        kubevirt.io/domain: vm-ubuntu03
    spec:
      architecture: amd64
      domain:
        devices:
          disks:
          - disk:
              bus: virtio
            name: root-disk
          - disk:
              bus: virtio
            name: cloudinitdisk
          rng: {}
          interfaces:
            - name: default
              masquerade: {}
      networks:
        - name: default
          pod: {}
      terminationGracePeriodSeconds: 0
      volumes:
      - name: root-disk
        persistentVolumeClaim:
          claimName: pvc-ubuntu03-bootdisk
      - name: cloudinitdisk
        cloudInitConfigDrive:
          userData: |-
            #cloud-config
            hostname: vm-ubuntu03
```
Subsequently, I update the CPU count in t1 to 2
```shell
# kubectl patch VirtualMachineInstancetype/t1 --type merge -p '{"spec":{"cpu":{"guest":2}}}'
```
After stopping the virtual machine and clearing the revisionName in the instanceType
```shell
# virtctl stop vm-ubuntu03
# kubectl patch vm/vm-ubuntu03 --type merge -p '{"spec":{"instancetype":{"revisionName":""}}}'
# vm start vm-ubuntu03
```
when I start the virtual machine again, I find that the `spec.domain.cpu.sockets` value changed to 2，but the  `Current CPU Topology` within the Virtual Machine Instance (VMI) has not been refreshed. 
```shell
Name:         vm-ubuntu03
Namespace:    default
Labels:       kubevirt.io/domain=vm-ubuntu03
              kubevirt.io/nodeName=k8s04
Annotations:  kubevirt.io/instancetype-name: t1
              kubevirt.io/latest-observed-api-version: v1
              kubevirt.io/storage-observed-api-version: v1
              kubevirt.io/vm-generation: 6
API Version:  kubevirt.io/v1
Kind:         VirtualMachineInstance
Metadata:
  Creation Timestamp:  2024-02-15T06:59:10Z
  Finalizers:
    kubevirt.io/virtualMachineControllerFinalize
    foregroundDeleteVirtualMachine
  Generation:  11
  Owner References:
    API Version:           kubevirt.io/v1
    Block Owner Deletion:  true
    Controller:            true
    Kind:                  VirtualMachine
    Name:                  vm-ubuntu03
    UID:                   8a8f523d-55d3-4905-be93-9409a6ced7fb
  Resource Version:        578510
  UID:                     93891038-44a3-4b18-a20f-66b3701f1593
Spec:
  Architecture:  amd64
  Domain:
    Cpu:
      Cores:    1
      Model:    host-passthrough
      Sockets:  2
      Threads:  1
    Devices:
      Disks:
        Disk:
          Bus:  virtio
        Name:   root-disk
        Disk:
          Bus:  virtio
        Name:   cloudinitdisk
      Interfaces:
        Masquerade:
        Name:  default
      Rng:
    Features:
      Acpi:
        Enabled:  true
    Firmware:
      Uuid:  2165306d-9e95-5673-a3d8-9cb65bcbaf33
    Machine:
      Type:  q35
    Memory:
      Guest:  1Gi
    Resources:
      Requests:
        Memory:       1Gi
  Eviction Strategy:  None
  Networks:
    Name:  default
    Pod:
  Node Selector:
    Cpu:                             true
  Termination Grace Period Seconds:  0
  Volumes:
    Name:  root-disk
    Persistent Volume Claim:
      Claim Name:  pvc-ubuntu03-bootdisk
    Cloud Init Config Drive:
      User Data:  #cloud-config
hostname: vm-ubuntu03
Status:
  Active Pods:
    2abd5637-bbb4-42c7-b465-45008f7c5079:  k8s04
  Conditions:
    Last Probe Time:       <nil>
    Last Transition Time:  2024-02-15T06:59:21Z
    Status:                True
    Type:                  Ready
    Last Probe Time:       <nil>
    Last Transition Time:  <nil>
    Status:                True
    Type:                  LiveMigratable
    Last Probe Time:       2024-02-15T07:00:09Z
    Last Transition Time:  <nil>
    Status:                True
    Type:                  AgentConnected
  Current CPU Topology:
    Cores:    1
    Sockets:  1
    Threads:  1
  Guest OS Info:
    Id:              ubuntu
...
```
**What you expected to happen**:
 the `Current CPU Topology` of the virtual machine should also change

**Environment**:
- KubeVirt version (use `virtctl version`): 1.1.1
- Kubernetes version (use `kubectl version`): 1.27.6
- VM or VMI specifications: N/A
- Cloud provider or hardware configuration: N/A
- OS (e.g. from /etc/os-release): N/A
- Kernel (e.g. `uname -a`): N/A
- Install tools: N/A
- Others: N/A

**Comment by 0xFelix** [comment]
Thanks for reporthing this @tghfly.

To me it looks like a bug in the hotplugging code. It is looking at the VM to fill in the `Current CPU Topology` status, which in case of instancetypes is not going to work, as the values are not set on the VM.

See here:
https://github.com/kubevirt/kubevirt/blob/05b862da023f71243f6b4510442b9fb2d98151dd/pkg/virt-controller/watch/vm.go#L1735

I suspect you will always see the same value of `1`, even if you create the VM with `t1` having two vCPUs from the start? (without changing the instancetype and restarting the VM in between)

**Comment by lyarwood** [comment]
https://github.com/kubevirt/kubevirt/pull/11455 and more specifically https://github.com/kubevirt/kubevirt/pull/11455/commits/a322edbd70146386e694773ef1ab17e7e9747164 will hopefully resolve this by moving this to the VMI webhook. I'll refresh that series now.

**Comment by lyarwood** [comment]
@tghfly this should be resolved in the upcoming `v1.3.0` release. Closing for now but feel free to reopen with a comment if there's still something to resolve here.

/close
